pmtiles.cpp

```diff
@@ -0,0 +1,418 @@
+#include <stdio.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <zlib.h>
+#include "pmtiles.hpp"
+#include <protozero/varint.hpp>
+#include <iostream>
+#include <sstream>
+#include "mvt.hpp"
+#include "write_json.hpp"
+#include "version.hpp"
+
+pmtilesv3 *pmtilesv3_open(const char *filename, char **argv, int force, const char* tmpdir) {
+	pmtilesv3 *outfile = new pmtilesv3;
+
+	struct stat st;
+	if (force) {
+		unlink(filename);
+	} else {
+		if (stat(filename, &st) == 0) {
+			fprintf(stderr, "%s: %s: file exists\n", argv[0], filename);
+			exit(EXIT_FAILURE);
+		}
+	}
+	outfile->ostream.open(filename,std::ios::out | std::ios::binary);
+
+	char tmpname[strlen(tmpdir) + strlen("/pmtiles.XXXXXX") + 1];
+	sprintf(tmpname, "%s%s", tmpdir, "/pmtiles.XXXXXX");
+	int tmpfd = mkstemp(tmpname);
+	close(tmpfd);
+
+	outfile->tmptilesname = tmpname;
+	outfile->tilestmp.open(tmpname,std::ios::out | std::ios::binary);
+	outfile->offset = 0;
+	return outfile;
+}
+
+void serialize_string(const std::string &str, std::stringstream &ss) {
+	size_t MAX_SIZE = 10;
+	uint8_t slen;
+	slen = str.size();
+	if (slen > MAX_SIZE) {
+		fprintf(stderr, "%s: header value exceeds limit\n", str.data());
+		exit(EXIT_FAILURE);
+	}
+	ss.write((char *)&slen,1);
+	ss.write(str.data(),slen);
+
+	char zero = 0;
+	for (size_t i = 0; i < MAX_SIZE-slen; i++) {
+    ss.write(&zero,sizeof(char));
+	}
+}
+
+// TODO: endianness check
```

Comment:
Should we bother with this - is tippecanoe ever going to be run on a big-endian machine?